### PR TITLE
Fixed bug in mixed path equality operator

### PR DESCRIPTION
**Pull request by 95DDB**
Previous code was working unexpectedely, as it results in an infinite recursive call; instead, due to some compiler bug probably, this wasn't happening and instead the call self == other was always returning true, even if the equality was indeed not satisfied.

### CODE REVIEW

File: primitives/src/merkle_tree/field_based_mht/path.rs [hunk 1]
```diff
@@ -375,6 +415,104 @@ impl<T: FieldBasedMerkleTreeParameters> PartialEq<FieldBasedMHTPath<T>>
...
-        self == other
+
+        // Paths must have the same length
+        if self.get_length() != other.get_length() {
+            return false;
+        }
+
+        // Iterate through both paths (NOTE: Could work also converting one of the two paths into
+        // the other type and then calling the corresponding Eq, but you save one iteration in this
+        // way)
```
Comment (nicholas-mainardi) [inline primitives/src/merkle_tree/field_based_mht/path.rs]: I would improve the clarity of the comment by saying that in this way you avoid sweeping twice the paths: saving one iteration seems a bit too generic

File: primitives/src/merkle_tree/field_based_mht/path.rs [hunk 2]
```diff
@@ -375,6 +415,104 @@ impl<T: FieldBasedMerkleTreeParameters> PartialEq<FieldBasedMHTPath<T>>
...
+        let binary_path2: TestBinaryPath = path2.clone().try_into().unwrap();
+        let binary_path3: TestBinaryPath = path3.clone().try_into().unwrap();
+        
+        assert_eq!(binary_path1, binary_path1);
+        assert_ne!(binary_path1, binary_path2);
+        assert_ne!(binary_path1, binary_path3);
+        assert_ne!(binary_path2, binary_path3);
+
+        // Mixed equality operator sanity checks
+        assert_eq!(binary_path1, path1);
```
Comment (nicholas-mainardi) [inline primitives/src/merkle_tree/field_based_mht/path.rs:511]: Just to rule out weird behaviors, I would add also a further check that inequality on equal path works (and similarly equality on different paths) by adding the following assertions:
`assert_eq!(binary_path2 != path2, false);`
`assert_eq!(binary_path1 == path2, false);`

**Review by nicholas-mainardi — APPROVED** [review]

File: primitives/src/merkle_tree/field_based_mht/path.rs (hunk 1 above)
Comment (95DDB) [reply primitives/src/merkle_tree/field_based_mht/path.rs]: Done

File: primitives/src/merkle_tree/field_based_mht/path.rs (hunk 2 above)
Comment (95DDB) [reply primitives/src/merkle_tree/field_based_mht/path.rs:511]: Done

**Review by nicholas-mainardi — APPROVED** [review]

**Review by Ken2889 — APPROVED** [review]

**Pull request merged**